### PR TITLE
Add ZIO.tupled

### PR DESCRIPTION
**Pull request by luis3m**
As suggested here https://github.com/zio/zio/issues/3564

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -3041,6 +3041,29 @@ object ZIO extends ZIOCompanionPlatformSpecific {
...
+  /**
+   * Sequentially zips the specified effects. Specialized version of mapN.
+   */
+  def tupled[R, E, A, B, C, D](
+    zio1: ZIO[R, E, A],
+    zio2: ZIO[R, E, B],
+    zio3: ZIO[R, E, C],
+    zio4: ZIO[R, E, D]
+  ): ZIO[R, E, (A, B, C, D)] =
+    mapN(zio1, zio2, zio3, zio4) { case tuple => tuple }
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala:3065]: You can do `mapN(zio1, zio2, zio3, zio4)((_, _, _, _))` here.

Comment (luis3m) [reply core/shared/src/main/scala/zio/ZIO.scala:3065]: @adamgfraser I will update in a few minutes

**Comment by luis3m** [comment]
@adamgfraser sure, I can add the parallel variants as well. Maybe open an issue about the naming conventions first? I can make those changes as well.

**Comment by adamgfraser** [comment]
@luis3m Sounds great!

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 1 above)
Comment (luis3m) [reply core/shared/src/main/scala/zio/ZIO.scala]: @adamgfraser Done

**Comment by luis3m** [comment]
@adamgfraser 

- Signed the CLA
- Updated partial functions
- Added parallel variants

**Pull request merged**